src/main/java/com/aws/greengrass/device/ClientDevicesAuthIPCAgent.java

```diff
@@ -0,0 +1,212 @@
+import java.security.cert.X509Certificate;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.function.Consumer;
+import javax.inject.Inject;
+
+import static com.aws.greengrass.ipc.common.ExceptionUtil.translateExceptions;
+
+/**
+ * Class to handle business logic for all client device auth component requests over IPC.
+ */
+public class ClientDevicesAuthIPCAgent {
+    private static final Logger logger = LogManager.getLogger(ClientDevicesAuthIPCAgent.class);
+    private static final String CERT_TYPE_SERVER = "SERVER";
+    private static final String CERT_TYPE_CLIENT = "CLIENT";
+
+    @Inject
+    @Setter(AccessLevel.PACKAGE)
+    ClientDevicesAuthService clientDevicesAuthService;
+    @Inject
+    @Setter(AccessLevel.PACKAGE)
+    CertificateManager certificateManager;
+
+    public SubscribeToCertificateUpdatesOperationHandler subscribeToCertificateUpdatesOperationHandler(
+            OperationContinuationHandlerContext context) {
+        return new SubscribeToCertificateUpdatesOperationHandler(context);
+    }
+
+    class SubscribeToCertificateUpdatesOperationHandler
+            extends GeneratedAbstractSubscribeToCertificateUpdatesOperationHandler {
+        private final String serviceName;
+        private final Consumer<X509Certificate> serverCertificateCallback = this::forwardServerCertificates;
+        private final Consumer<X509Certificate[]> clientCertificateCallback = this::forwardClientCertificates;
+        private String csr;
+
+        protected SubscribeToCertificateUpdatesOperationHandler(OperationContinuationHandlerContext context) {
+            super(context);
+            serviceName = context.getAuthenticationData().getIdentityLabel();
+        }
+
+        @Override
+        protected void onStreamClosed() {
+
+        }
+
+        @SuppressWarnings("PMD.PreserveStackTrace")
+        @Override
+        public SubscribeToCertificateUpdatesResponse handleRequest(
+                SubscribeToCertificateUpdatesRequest subscribeToCertificateUpdatesRequest) {
+            return translateExceptions(() -> {
+                try {
+                    clientDevicesAuthService.doAuthorization(serviceName);
+                } catch (AuthorizationException e) {
+                    throw new UnauthorizedError(e.getMessage());
+                }
+                byte[] publicKey = subscribeToCertificateUpdatesRequest.getEncryptionKeys().getPublicKey();
+                byte[] privateKey = subscribeToCertificateUpdatesRequest.getEncryptionKeys().getPrivateKey();
+                String algorithm =
+                        subscribeToCertificateUpdatesRequest.getEncryptionKeyOptions().getKeyAlgorithmAsString();
+                try {
+                    csr = getCSR(publicKey, privateKey, algorithm);
```

Comment:
I know the current cert generation code expects a CSR, but we can probably avoid this altogether.

Ok for now, but keep that in the back of your mind to see whether we can get rid of some of that code